File: problem3/NestedListWeightSum.java
// Time Complexity : O(n), n -> Number of integers
// Space Complexity : O(d), d -> Maximum level of nesting
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No
package problem3;

import java.util.List;

public class NestedListWeightSum {
	public int depthSum(List<NestedInteger> nestedList) {
		return helper(nestedList, 1);
	}

	public int helper(List<NestedInteger> nestedList, int depth) {
		int sum = 0;
		for (NestedInteger num : nestedList) {
			if (num.isInteger()) {
				sum += num.getInteger() * depth;
			} else {
				sum += helper(num.getList(), depth + 1);
			}
		}
		return sum;
	}

	public static void main(String[] args) {
//		NestedListWeightSum obj = new NestedListWeightSum();

	}

}

File: nested_list_weight_sum.py
class Solution:
    def depthSum(self, nestedList: List[NestedInteger]) -> int:
        
        ## apprach 2 : bfs
        ## T.C = O(n)
        ## S.C = O(n)
        sm = 0
        q = [nestedList]
        depth = 1
        
        while q:

            for i in range(len(q)):
                elems = q.pop(0)
                for el in elems:     
                    if el.isInteger():
                        sm += el.getInteger() * depth
                    else:
                        q.append(el.getList())
            depth += 1

        return sm
        
        ## approach 1 : dfs
        ## T.C = O(n)
        ## S.C = O(depth)
        sm = 0
        def dfs(nums, depth):
            nonlocal sm

            for el in nums:
                if el.isInteger():
                    sm += el.getInteger() * depth
                else:
                    dfs(el.getList(), depth + 1)
        
        dfs(nestedList, 1)
        return sm
